libs/SalesforceSDK/src/com/salesforce/androidsdk/ui/OAuthWebviewHelper.java

```diff
@@ -541,8 +569,37 @@ public boolean shouldOverrideUrlLoading(WebView view, WebResourceRequest request
                         onAuthFlowComplete(tr);
                     }
                 }
+            } else {
+                // Only allow redirect for requests with authorization headers for Salesforce urls.
+                if (request.getRequestHeaders() != null && request.getRequestHeaders().containsKey("Authorization")) {
+                   return isSalesforceUrl(formattedUrl);
+                }
             }
-            return isDone;
+
+            return authFlowFinished;
+        }
+
+        // List from https://help.salesforce.com/s/articleView?language=en_US&id=sf.domain_name_url_formats.htm&type=5
+        private boolean isSalesforceUrl(String host) {
+            return (host != null && (host.endsWith(".salesforce.com") ||
+                    host.endsWith(".force.com") ||
+                    host.endsWith(".sfdcopens.com") ||
+                    host.endsWith(".site.com") ||
+                    host.endsWith(".lightning.com") ||
+                    host.endsWith(".salesforce-sites.com") ||
+                    host.endsWith(".force-user-content.com") ||
+                    host.endsWith(".salesforce-experience.com") ||
+                    host.endsWith(".salesforce-scrt.com")));
+        }
+
+        private boolean isNewLoginUrl(Uri uri) {
+            if (!uri.toString().contains(getLoginUrl()) && isSalesforceUrl(uri.getHost())) {
+                String path = uri.getQuery();
+                return path != null &&
+                        path.startsWith("startURL=/setup/secur/RemoteAccessAuthorizationPage.apexp?source=");
```

Comment:
This condition may be too specific as it runs the chance of breaking if the url changes in the future.